### PR TITLE
Fix socket poll and handling in windows

### PR DESCRIPTION
**Pull request by PabloMK7**
This branch fixes three aspects of socket handling in `soc:u`:

1) In windows, the socket handle is actually 64 bits, but was being passed to the 3DS side as a 32 bit value. Normally this is not a problem because windows socket handles are in the 32 bit range, but it's better to stay safe if some weird change uses the upper 32 bits.

2) In the poll function, the std::transform used to copy from different arrays doesn't guarantee the order of its elements (which could then be passed to the 3ds in the wrong order). A for loop is used instead.

3) On Linux, invalid poll flags in the input field are ignored, while in windows the poll functions return `EINVAL` if there are any invalid input fields. The changes make sure to remove those flags on windows if they appear in the input (for example, `POLLPRI` is unsupported on windows). This caused homebrew apps using libcurl to not work on windows, this PR fixes that.

Commit [c80acc8](https://github.com/citra-emu/citra/pull/6166/commits/c80acc82b8d8e2fc51340b588f843cebbf6cc781) fixes another problem:

4) Blocking operations such as `connect` or `poll` block the entire host, including the guest timer. This causes the guest to "see" that the blocking operation was completed immediately, breaking any kind of timeout checks. Since I'm pretty sure other services will face a similar problem, I've created two methods in the Timer class, which help asserting that there aren't stray `StartAdjust` or `EndAdjust` calls.  The soc module now uses them to adjust the guest timer depending on how long the blocking operation took.

- - -
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/6166)

### CODE REVIEW

File: src/core/hle/service/soc_u.cpp [hunk 1]
```diff
@@ -700,21 +778,32 @@ void SOC_U::Poll(Kernel::HLERequestContext& ctx) {
 
     // The 3ds_pollfd and the pollfd structures may be different (Windows/Linux have different
     // sizes)
-    // so we have to copy the data
+    // so we have to copy the data in order
     std::vector<pollfd> platform_pollfd(nfds);
-    std::transform(ctr_fds.begin(), ctr_fds.end(), platform_pollfd.begin(), CTRPollFD::ToPlatform);
+    for (u32 i = 0; i < nfds; i++) {
+        platform_pollfd[i] = std::move(CTRPollFD::ToPlatform(*this, ctr_fds[i]));
```
Comment (Morph1984) [inline src/core/hle/service/soc_u.cpp:784]: std::move is pessimizing here

File: src/core/hle/service/soc_u.cpp [hunk 2]
```diff
@@ -700,21 +778,32 @@ void SOC_U::Poll(Kernel::HLERequestContext& ctx) {
...
+    PreTimerAdjust();
     s32 ret = ::poll(platform_pollfd.data(), nfds, timeout);
+    PostTimerAdjust();
 
-    // Now update the output pollfd structure
-    std::transform(platform_pollfd.begin(), platform_pollfd.end(), ctr_fds.begin(),
-                   CTRPollFD::FromPlatform);
+    // Now update the output 3ds_pollfd structure
+    for (u32 i = 0; i < nfds; i++) {
+        ctr_fds[i] = std::move(CTRPollFD::FromPlatform(*this, platform_pollfd[i]));
```
Comment (Morph1984) [inline src/core/hle/service/soc_u.cpp:793]: ditto

File: src/core/core_timing.h [hunk 3]
```diff
@@ -227,6 +232,9 @@ class Timing {
         s64 downcount = MAX_SLICE_LENGTH;
         s64 executed_ticks = 0;
         u64 idled_cycles = 0;
+
+        std::chrono::time_point<std::chrono::high_resolution_clock> adjust_value_last;
```
[the remark this answers is not in the document]
Comment (PabloMK7) [reply src/core/core_timing.h]: Indeed, fixed

File: src/core/hle/service/soc_u.cpp (hunk 1 above)
Comment (PabloMK7) [reply src/core/hle/service/soc_u.cpp]: Fixed!

File: src/core/hle/service/soc_u.cpp (hunk 2 above)
Comment (PabloMK7) [reply src/core/hle/service/soc_u.cpp]: ditto

**Comment by SachinVin** [comment]
How would timeout checks work if the ticks are updated after the blocking operation is completed, from the applications perspective the op has returned a valid (success/fail) result, would there be any need for a timeout then?
Kind of a related question, if the app is running at say 2x speed, the in-app timer would timeout in half the time vs the actual clock.

**Comment by PabloMK7** [comment]
Imagine the following pseudocode on the 3ds:
```
totalTicks = 0;
while (true) {
        prevTicks = svcGetSystemTick();
        socketPoll(timeout of 15 milliseconds)
        currTicks = svcGetSystemTick();
        
        totalTicks = totalTicks + (currTicks - prevTicks);
        if (totalTicks > 15 seconds)
            return has timed out;
}
```
Normally on the 3ds, because the soc module just runs in another process, the system tick continue increasing while the game is blocked on `socketPoll`. If we specify a timeout of 15 milliseconds, then the game will see `currTicks - prevTicks = 15 milliseconds + time of running the instructions between svc`. However on citra, when we call `::poll`, the entire emulator thread is paused, including the system tick. That causes the game to see `currTicks - prevTicks = 0 milliseconds + time running the instructions between svc`, so the timeout calculations fail (and the timeout turns into a huge size). That's why we have to manulayy calculate how many ticks the 3ds would have taken in socket blocking operations, and advance the system tick accordingly.

You are right about the emulation speed, `EndAdjust` needs to multiply the calculated ticks by the core speed multiplier.

**Comment by SachinVin** [comment]
Thanks, for the detailed explanation, I get it now.

**Pull request merged**